### DISCUSSION TITLE
module 'djot' not found:

### DISCUSSION THREAD

**Discussion by the-solipsist** [jgm/djot#149] [Q&A] [answered]
I can't seem to use `djot-reader.lua` with `pandoc` (whereas `djot-writer.lua` seems to work fine).

```
$ cd ~/.local/share/pandoc/filters
$ wget https://raw.githubusercontent.com/jgm/djot/main/djot-reader.lua -q
$ wget https://raw.githubusercontent.com/jgm/djot/main/djot.lua -q
$ wget https://raw.githubusercontent.com/jgm/djot/main/djot-writer.lua -q
$ cd ~/documents
$ pandoc -f ~/.local/share/pandoc/filters/djot-reader.lua mpy-001.md -t html
Error running Lua:
/home/sol/.local/share/pandoc/filters/djot-reader.lua:1: module 'djot' not found:
	no field package.preload['djot']
	no file '/usr/local/share/lua/5.4/djot.lua'
	no file '/usr/local/share/lua/5.4/djot/init.lua'
	no file '/usr/local/lib/lua/5.4/djot.lua'
	no file '/usr/local/lib/lua/5.4/djot/init.lua'
	no file './djot.lua'
	no file './djot/init.lua'
	no file '/usr/local/lib/lua/5.4/djot.so'
	no file '/usr/local/lib/lua/5.4/loadall.so'
	no file './djot.so'
stack traceback:
	/home/sol/.local/share/pandoc/filters/djot-reader.lua:1: in main chunk
$ sudo luarocks install djot
Warning: Failed searching manifest: Failed downloading https://luarocks.org/manifest-5.3 - host or service not provided, or not known
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/djot-0.1.0-1.src.rock

djot 0.1.0-1 depends on lua >= 5.1 (5.3-1 provided by VM)
djot 0.1.0-1 is now installed in /usr/local (license: MIT)

$ pandoc -f ~/.local/share/pandoc/filters/djot-reader.lua mpy-001.md -t html
Error running Lua:
/home/sol/.local/share/pandoc/filters/djot-reader.lua:1: module 'djot' not found:
	no field package.preload['djot']
	no file '/usr/local/share/lua/5.4/djot.lua'
	no file '/usr/local/share/lua/5.4/djot/init.lua'
	no file '/usr/local/lib/lua/5.4/djot.lua'
	no file '/usr/local/lib/lua/5.4/djot/init.lua'
	no file './djot.lua'
	no file './djot/init.lua'
	no file '/usr/local/lib/lua/5.4/djot.so'
	no file '/usr/local/lib/lua/5.4/loadall.so'
	no file './djot.so'
stack traceback:
	/home/sol/.local/share/pandoc/filters/djot-reader.lua:1: in main chunk
$ wget https://raw.githubusercontent.com/jgm/djot/main/djot.lua -q
$ pandoc -f ~/.local/share/pandoc/filters/djot-reader.lua mpy-001.md -t html
Error running Lua:
./djot.lua:2: module 'djot.block' not found:
	no field package.preload['djot.block']
	no file '/usr/local/share/lua/5.4/djot/block.lua'
	no file '/usr/local/share/lua/5.4/djot/block/init.lua'
	no file '/usr/local/lib/lua/5.4/djot/block.lua'
	no file '/usr/local/lib/lua/5.4/djot/block/init.lua'
	no file './djot/block.lua'
	no file './djot/block/init.lua'
	no file '/usr/local/lib/lua/5.4/djot/block.so'
	no file '/usr/local/lib/lua/5.4/loadall.so'
	no file './djot/block.so'
	no file '/usr/local/lib/lua/5.4/djot.so'
	no file '/usr/local/lib/lua/5.4/loadall.so'
	no file './djot.so'
stack traceback:
	./djot.lua:2: in main chunk
	[C]: in function 'require'
	/home/sol/.local/share/pandoc/filters/djot-reader.lua:1: in main chunk

```

What am I doing wrong?

**Comment by jgm** [comment, marked as answer]
You'll need to install the djot package via luarocks:

```
make install
```

from the djot repository should do this. Or, you can use djot-reader.lua in the djot source directory, and add to the top of djot-reader.lua
```
package.path = './?.lua;' .. package.path
```
Note: the API is a bit unstable and it's possible that djot-reader.lua may stop working if I change the djot API. If that happens, remind me and I can fix it.

**Reply by the-solipsist** [reply]
I'm a bit confused. Am I not installing the djot package via luarocks when I do `sudo luarocks install djot`?  (I'd tried `luarocks install djot --local` as well, but it didn't seem to make a difference.)

So, I should `git clone` the `djot` repo into `~/.local/share/pandoc/filters/djot/`, and then run `make install` from that directory?

Also, I tried adding `package.path = './?.lua;' .. package.path` as the first line in `djot-reader.lua`, but that didn't seem to make a difference.

**Reply by jgm** [reply]
That should install it, yes, but maybe your LUA_PATH environment variable is set wrong?
The error message you got shows the path it is searching,  so look in those locations.  Maybe your system put the rocks somewhere else?

> Also, I tried adding package.path = './?.lua;' .. package.path as the first line in djot-reader.lua, but that didn't seem to make a difference.

Are you running it inside the djot source directory (i.e. the directory cloned from the git repository)?

**Comment by bpj** [comment]
There are several env vars which need/can be set. See https://web.archive.org/web/20210127030419/http://www.thijsschreijer.nl/blog/?p=1025

**Comment by tbdalgaard** [comment]
I have the same issue. From the documentation I was told to copy the reader and writer filters to the working directory and use them from there when using Pandoc.
This got me different errors so I gave up and are now running this via the repo directory which I hoped I could have avoided, but if it works, then at least that's better than nothing.

Could something be done to reference how to install these filters into Pandoc?

I am running Mac OS 10.15.7 here, but I think the old OS is not the root of the problem here.

**Reply by jgm** [reply]
You need to install the djot library using luarocks.
Since I've now uploaded a new version to the rocks repository, it should work just to do `luarocks install djot`.  (First you'll need to get luarocks, of course, which you can do through homebrew.) If it still doesn't work, then the problem is that your LUA_PATH is not right.

**Reply by tbdalgaard** [reply]
To be more specific here is what I get when trying to run the custom reader and writer for Pandoc from the repository directory:
```
Error running Lua:
djot-reader.lua:76: attempt to index a nil value (local 'node')
stack traceback:
	djot-reader.lua:263: in function <djot-reader.lua:261>
	(...tail calls...)
	djot-reader.lua:98: in method 'render_children'
	djot-reader.lua:67: in function <djot-reader.lua:66>
	(...tail calls...)
	djot-reader.lua:98: in method 'render_children'
	djot-reader.lua:117: in function <djot-reader.lua:114>
	(...tail calls...)
	djot-reader.lua:98: in method 'render_children'
	djot-reader.lua:117: in function <djot-reader.lua:114>
	(...tail calls...)
	djot-reader.lua:98: in method 'render_children'
	djot-reader.lua:111: in function <djot-reader.lua:108>
	(...tail calls...)

```

How can I fix this?

**Comment by jgm** [comment]
I've found a way to avoid these problems by creating a single lua file that includes the custom reader or writer and the whole djot library.  First you need to install the "amalg" tool: `luarocks install amalg`.  Then go to the djot repository directory and do this:
```
LUA_PATH="./?.lua;" amalg.lua djot djot.ast djot.block djot.filter djot.inline djot.attributes djot.emoji djot.html djot.json -s djot-reader.lua -o djot-reader-combined.lua
```
You should now have a single file `djot-reader-combined.lua` which you can move anywhere you like and use as the custom reader.  (For the writer, same thing except `s/reader/writer/`.)

**Reply by tbdalgaard** [reply]
Seems I was only lucky once, so now I can't get it to work again. Have tried to reinstall lua, luarocks, amalk and djot itself, but nothing seems to work probably now.
Am using Mac OS 10.15-7 which is by far not the latest OS, but I hoped this would have worked here as well.

**Comment by tbdalgaard** [comment]
Nice! That works great here.
Will this be integrated somehow into the filters that are provided or can the documentation be expanded?

I really enjoy writing in Djot, and find it very easy to use as Markdown, but I much prefer Djot over Markdown in the long run.

**Comment by jgm** [comment]
I've added some documentation on this and a handy Makefile target.
I'll also start testing the custom reader/writer on CI, so I'll know when they break.

**Reply by the-solipsist** [reply]
Thanks, the added text in the readme is most helpful.

What finally worked for me (using fish shell):
```
$ luarocks path
'/usr/local/share/lua/5.3/?.lua;/usr/local/share/lua/5.3/?/init.lua;/usr/local/lib/lua/5.3/?.lua;/usr/local/lib/lua/5.3/?/init.lua;/usr/share/lua/5.3/?.lua;/usr/share/lua/5.3/?/init.lua;./?.lua;./?/init.lua;/home/sol/.luarocks/share/lua/5.3/?.lua;/home/sol/.luarocks/share/lua/5.3/?/init.lua'
export LUA_CPATH='/usr/local/lib/lua/5.3/?.so;/usr/lib/x86_64-linux-gnu/lua/5.3/?.so;/usr/lib/lua/5.3/?.so;/usr/local/lib/lua/5.3/loadall.so;./?.so;/home/sol/.luarocks/lib/lua/5.3/?.so'
export PATH=...

$ set -Ux --path LUA_PATH '/usr/local/share/lua/5.3/?.lua;/usr/local/share/lua/5.3/?/init.lua;/usr/local/lib/lua/5.3/?.lua;/usr/local/lib/lua/5.3/?/init.lua;/usr/sh
are/lua/5.3/?.lua;/usr/share/lua/5.3/?/init.lua;./?.lua;./?/init.lua;/home/sol/.luarocks/share/lua/5.3/?.lua;/home/sol/.luarocks/share/lua/5.3/?/init.lua'

$ pandoc --from ~/.local/share/pandoc/filters/djot-reader.lua mpy-001.dj --to html
```

**Comment by tbdalgaard** [comment]
Thanks, but I still get an error. Have pasted from below after following the instructions in readme.md.
How can I fix this?

```
Error running Lua:
...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3965: bad argument #1 to 'gsub' (string expected, got nil)
stack traceback:
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3965: in function <...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3950>
	(...tail calls...)
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3840: in method 'render_children'
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3859: in function <...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3856>
	(...tail calls...)
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3840: in method 'render_children'
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3859: in function <...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3856>
	(...tail calls...)
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3840: in method 'render_children'
	...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3853: in function <...rs/tbd/.local/share/pandoc/readers/djot-reader.amalg.lua:3850>
	(...tail calls...)

```

Sendt fra min Mac Mini via Apple Mail

**Comment by jgm** [comment]
I think that was a bug I fixed this morning. Do you have the latest code from github?

**Comment by tbdalgaard** [comment]
Yes, I cloned this just less than two hours ago, and it still breaks.
After doing a totally new clone of the repository I managed to create the amalg.lua files, but if I copy them into my data-dir in Pandoc under readers and writers the reader will not function.
I have tried to reinstall everything and installing Djot from the source provided via the cloned repository.

I use the GitHub for desktop on Mac OS here if that can be of any help.

**Comment by tbdalgaard** [comment]
This is getting more and more interesting.
The read and write works fine if all stays in the djot repository directory including the djot-reader.amalg.lua and djot-writer.amalg.lua files. If I copy those to the data-dir for Pandoc and put djot-reader.amalg.lua in the subdirectory: readers
and the djot-writer.amalg.lua in the subdirectory: writers
it all starts to break.



Sendt fra min Mac Mini via Apple Mail

**Reply by jgm** [reply]
If you're using the latest dev pandoc, there's a recent change that wants you to put custom readers/writers in "custom" rather than "readers".

**Comment by tbdalgaard** [comment]
I only use the official version op Pandoc, and the issue is present every time I try to use the scripts outside the djot repository directory. That includes the scripts, but also the files I try to work with. How can this be fixed?I ended installing Djot via themake installfile from the repository. Sendt fra min iPhone via Braille Sense U2/MBrailleDen 30. dec. 2022 kl. 01.06 skrev John MacFarlane ***@***.***>:﻿
If you're using the latest dev pandoc, there's a recent change that wants you to put custom readers/writers in "custom" rather than "readers".

**Reply by jgm** [reply]
Sorry, I don't know how to help you. It worked for me (in `readers`) with the last official release.  Make sure you're using the right user data directory (see `pandoc --version`).

**Comment by tbdalgaard** [comment]
Ok will try again when I am back home.Perhaps there is something going on while I tried reinstalling all the packages. Hope you don't mind I update this, if I find something. Sendt fra min iPhone via Braille Sense U2/MBrailleDen 31. dec. 2022 kl. 00.11 skrev John MacFarlane ***@***.***>:﻿
Sorry, I don't know how to help you. It worked for me (in readers) with the last official release.  Make sure you're using the right user data directory (see pandoc --version).

**Comment by tbdalgaard** [comment]
I now tried to clean up from all the install I have done in the past, and ran the pandoc from the DJOT-directory that I cloned from GitHub.

´´´

I used this:
pandoc -f djot-reader.amalg.lua -t djot-writer.amalg.lua -o test-djot.txt README.MD

Got the following output:
Error running Lua:
djot-reader.amalg.lua:3965: bad argument #1 to 'gsub' (string expected, got nil)
stack traceback:
	djot-reader.amalg.lua:3965: in function <djot-reader.amalg.lua:3950>
	(...tail calls...)
	djot-reader.amalg.lua:3840: in method 'render_children'
	djot-reader.amalg.lua:3859: in function <djot-reader.amalg.lua:3856>
	(...tail calls...)
	djot-reader.amalg.lua:3840: in method 'render_children'
	djot-reader.amalg.lua:3859: in function <djot-reader.amalg.lua:3856>
	(...tail calls...)
	djot-reader.amalg.lua:3840: in method 'render_children'
	djot-reader.amalg.lua:3853: in function <djot-reader.amalg.lua:3850>
	(...tail calls...)

´´´
My data-dir is: ~/.local/share/pandoc
and I have saved the files in both readers and writers directories for the respective files so djot-reader goes into readers and djot-writer goes into writers.
Have tried installing Pandoc from both the Homebrew and the installer, and ended trying with the installer again.
I have installed DJot by typing:

´´´
brew install luarocks

luarocks install djot

´´´
All of this seems to work fine. Then I do:

´´´
make djot-reader.amalg.lua
make djot-writer.amalg.lua
´´´
It all looks fine until I run the Pandoc command.

**Reply by jgm** [reply]
To be safe, try (in the djot repository):
```
git pull
make -B djot-reader.amalg.lua
make -B djot-writer.amalg.lua
```

**Comment by tbdalgaard** [comment]
Hi!

When I use the
´´´

luarocks install djot

I get the following error:
djot 0.2.0-1 depends on lua >= 5.1 (5.4-1 provided by VM)
mkdir: /usr/local/lib/luarocks/rocks-5.4/djot/0.2.0-1/lua: Permission denied
Error: failed making directory /usr/local/lib/luarocks/rocks-5.4/djot/0.2.0-1/lua

I also just reinstalled Xcode developer tools again.

´´´

Sendt fra min Mac Mini via Apple Mail

**Reply by jgm** [reply]
OK, then
```
luarocks install --local djot
```
This will put djot in your home directory. You'll need to make sure your LUA_PATH is set correctly. See the luarocks documentation.
```
luarocks path
```
will give you the commands you need to set it properly.

**Comment by tbdalgaard** [comment]
Thanks for the advice, but the documentation for Luarocks is a little technical, and I may have done something wrong during the process.
I get the same error, so I am out of time and luck at the moment.

I really hoped I could have used some of my vacation time playing with Djot, as I really wanted to. Instead I have lost hours trying to fix something that broke because I thought it was a good idea to update Djot.
I hope this installation gets easier and that djot and Pandoc will be easier to pair for the future.
When I got Djot working I was very happy, and I really was looking forward to write and use it as my primary format of choice here, but due to the technical issues I have faced I may need to pause this until it gets a little more user friendly to install.
I may have failed this since I tried installing Luarocks via Homebrew and that installed Lua as well.

I will try to look into this again later in 2023, since I **love** the ideas behind Djot and was hoping that I could start this year adopting it as my primary format for composing documents.

Thanks for your help, you or any other can share some light of why this upgrade broke so much on my system and especially how to fix this, I will much appreciate it.
I think the reader is the main issue, and strangely enough not the writer on my system.


Sendt fra min Mac Mini via Apple Mail

**Reply by bpj** [reply]
[Here][] is a good introduction to Lua path. The original URL gives a 404 but it is preserved on the wayback machine

[Here]: https://web.archive.org/web/20210127030419/http://www.thijsschreijer.nl/blog/?p=1025

**Comment by jgm** [comment]
You can try my new JavaScript implementation at <https://github.com/jgm/djot.js>.
You'll need node to install it, but that is easy to get.
It provides a command-line tool that can convert between djot and pandoc JSON.
See the README.
